Jackie transaction fee

Yingjie-Zhang-dev opened the pull request:
Change the transaction fee to 5 basic points.

In addition, I've found that when we call the withdraw function to withdraw money, the unit is Wei, while when we call the donate function to donate money, the unit is ETH. 
```solidity
function withdraw(uint256 _amount) public payable onlyAdmin {
        require(address(this).balance > _amount, "There is no Ether to withdraw");
        outgoing.push(Transaction(msg.sender, _amount, block.timestamp, block.number, tx.gasprice, 2));

        // use the transfer method to transfer the amount to the admin's address
        (bool success, ) = admin.call{value: _amount}("");
        require(success, "Withdrawal failed");
    }
``` 
This makes that after we donate some ETH, we have to input 18 zeros in order to withdraw 1 ETH.
If we set the withdraw "_amount" to "_amount * 1e18", then the withdraw will be at least 1 ETH, if our contract only has less than 1 ETH balance, then the withdraw will fail.

Should we handle this problem? And how should we deal with the decimals?

Yingjie-Zhang-dev committed:
transaction fee

Change transaction fee to 5 basic points.

Yingjie-Zhang-dev committed:
Fix error

finnformica reviewed (commented):
I would move the 5bp out into its own variable - feeBasisPoints, or something - to make it easier for us to change in the future and for someone looking at our code to understand what the number is.

You've also raised a great point and I think we should handle it as you've suggested - would be easy to just define `amount = _amount * 1e18` and then using `amount` in the code instead.

Yingjie-Zhang-dev committed:
upload CauseContract.sol

1. move the basic points to its own variable so that it's easy to be changed
2. change the amount to amount * 1e18 in function (withdraw), unit: Wei-> ETH

Yingjie-Zhang-dev committed:
Delete unwanted comments

Yingjie-Zhang-dev commented:
> I would move the 5bp out into its own variable - feeBasisPoints, or something - to make it easier for us to change in the future and for someone looking at our code to understand what the number is.
> 
> You've also raised a great point and I think we should handle it as you've suggested - would be easy to just define `amount = _amount * 1e18` and then using `amount` in the code instead.

Thanks for your helpful advice! I've modified the codes as you suggested, and it's much more readable now. Moving the 5bp out into its own variable and defining `amount` have definitely made it easier to understand and maintain.

finnformica committed:
increase fee from 0.05% to 0.5%

finnformica reviewed (approved):
Just added a commit increasing the fee from 5bps to 50bps which is 0.05% to 0.5% - all looks good though

Yingjie-Zhang-dev committed:
add function checkIfIdUnique

Check whether the ID is unique or will throw an error.

Yingjie-Zhang-dev committed:
show the transaction fee

Make it possible to visually see the transaction fee after making a donation.

Yingjie-Zhang-dev commented on CauseContract.sol:
```diff
@@ -52,20 +53,25 @@ contract CauseContract {
 
-        blockchange.transfer(msg.value * feePercent / 100);
+        blockchange.transfer(transactionFee);
+        return transactionFee; // Fee in Wei
```
Since Solidity doesn't deal with decimals, I find it hard to convert the unit from Wei to ETH here. 
I believe the unit conversion may be handled quickly at the front end.

finnformica replied:
Yup I agree that will make it easier, I'll merge this PR for our new sprint tomorrow
